feat(database): Database Filtering via custom configuration

Antonio-RiveroMartnez opened the pull request:
### SUMMARY
There might be scenarios where you want to perform custom filtering on the list of databases returned by the DatabaseRestApi, right now, there's no way besides monkey patching that you can do so. This PR adds a new config definition in our app and makes use of it in the `DatabaseFilter` which is applied to all searches so you can add custom filtering if needed via config, adding live filtering capabilities to our searches and easing customization.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
<!--- Skip this if not applicable -->

### TESTING INSTRUCTIONS
1. Add a filtering function in your config file so all GET or GET list requests can make use of it and return a filtered result

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [X] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
Database Filtering:

- Add new configuration so we can inject extra filters to our databases when running the DatabaseFilter in base_filters
- Add tests for our new config and its usage

hughhhh commented on superset/databases/filters.py:
```diff
@@ -41,6 +41,16 @@ class DatabaseFilter(BaseFilter):  # pylint: disable=too-few-public-methods
+        # filtering.
+
+        if dynamic_filter := current_app.config["EXTRA_DYNAMIC_DATABASE_FILTER"]:
+            query = dynamic_filter(query)
```
what happens when the `dynamic_filter` isn't defined? looking at the code it seems like ` return query.filter(` will fail.

Can we add a condition to check if the `dynamic_filter` is defined, then have a test to back it

Antonio-RiveroMartnez replied:
Hmm how so? `dynamic_filter` is only defined if the `EXTRA_DYNAMIC_DATABASE_FILTER ` is defined in the config and if so, it will apply it to the query, otherwise the code is never going to enter into this conditional.

Antonio-RiveroMartnez commented on tests/integration_tests/databases/api_tests.py:
```diff
@@ -3632,3 +3632,77 @@ def test_validate_sql_endpoint_failure(self, get_validator_by_name):
+        uri = f"api/v1/database/"
+        rv = self.client.get(uri)
+        data = json.loads(rv.data.decode("utf-8"))
+        self.assertEqual(data["count"], len(dbs))
```
@hughhhh here I am testing our current behavior (default) where all databases must be returned if nothing is being set in the config, so `dynamic_filter` is not defined. Then, I'm adding the patch for the config to add the filter function and testing it's being applied because `dynamic_filter` is defined.

Antonio-RiveroMartnez replied:
Added explicit assertions to check whether the filter method has been called when defined.

betodealmeida replied:
> here I am testing our current behavior (default) where all databases must be returned if nothing is being set in the config, so dynamic_filter is not defined. Then, I'm adding the patch for the config to add the filter function and testing it's being applied because dynamic_filter is defined.

Can you write that down in the function docstring? :)

john-bodley commented on superset/databases/filters.py:
```diff
@@ -41,6 +41,16 @@ class DatabaseFilter(BaseFilter):  # pylint: disable=too-few-public-methods
     # TODO(bogdan): consider caching.
 
     def apply(self, query: Query, value: Any) -> Query:
+        # Dynamic Filters need to be applied to the Query before we filter
```
Feels like a good place for a docstring comment.

john-bodley reviewed (commented):
Could you provide some scenarios where you would use this? I also wonder if this should be generalized for all entity types.

Additionally I’m not saying I’m a supporter of Monkey Patching, but we should be cognizant that sometimes the method is used to relax (as opposed to further restrict) filters whereas this approach only addresses the later.

Antonio-RiveroMartnez committed:
Database filtering:

- Add explicit assertions to validate the filter is not called when not defined
- Use docstring comment

Antonio-RiveroMartnez committed:
Database filtering:

- Generalize the new config so we can use it for other entities down the road.
- Pull the new database key from the config so we apply any given filter method in the databases API
- Adjust our tests with the new config name and structure

betodealmeida reviewed (approved):
Elegant!

Antonio-RiveroMartnez committed:
Database Filtering:

- Add more info to our comments in our tests

john-bodley commented:
@Antonio-RiveroMartnez regarding your comment, 

> There might be cases where I have a given database that I want to make visible/hidden to my users based on a Feature Flag (live change).

If that was the case shouldn't the permission override logic be handled in the security manager? Note I'm not blocking this change, I just think there's merit in making sure we measure twice cut once in terms of adding new functionality.
